### PR TITLE
feat(zero-bubble): support zero bubble pipeline parallism

### PR DESCRIPTION
**Pull request by ChengYao-amd**

### CODE REVIEW

File: primus/backends/megatron/core/distributed/finalize_model_grad.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+from typing import List, Optional
```
Comment (wenxie-amd) [inline primus/backends/megatron/core/distributed/finalize_model_grad.py:10]: Add a license header to every Python script. Please also include ZeroBubble if we reference any code from it.

File: primus/backends/megatron/core/extensions/primus_turbo.py [hunk 2]
```diff
@@ -406,10 +416,22 @@ def __init__(
         tp_comm_buffer_name: Optional[str] = None,
         tp_group: Optional[torch.distributed.ProcessGroup] = None,
     ):
+        print("debug init turbo linear")
```
Comment (wenxie-amd) [inline primus/backends/megatron/core/extensions/primus_turbo.py]: Use the logger of primus

**Review by wenxie-amd — DISMISSED** [review]

File: primus/backends/megatron/core/transformer/transformer_layer.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
+###############################################################################
+# Some parts of this code are copied and modified from
+# Sea AI Lab's zero-bubble-pipeline-parallelism project
+# (https://github.com/sail-sg/zero-bubble-pipeline-parallelism).
+#
+# Modification Copyright© 2025 Advanced Micro Devices, Inc. All rights reserved.
+# See LICENSE for license information.
+###############################################################################
+
+# Copyright (c) 2025, NVIDIA CORPORATION. All rights reserved.
```
Comment (lhzhang333) [inline primus/backends/megatron/core/transformer/transformer_layer.py]: should be removed?

File: primus/backends/megatron/core/pipeline_parallel/zerobubble/scheduler/zbv.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+            end_time[_id] = _tmp
+            cur_time[stage] = _tmp
+            mem[stage] += self.fbw_mem[cat]
+            # noinspection PyTypeChecker
+            schedule[stage].append((cat, chunk, _cnt))
+            if cat == 1:
+                pending_w[stage].append((2, chunk, _cnt))
+            count[stage][cat * 2 + chunk] += 1
+
+        # for _ in range(2 * self.n_stage):
```
Comment (lhzhang333) [inline primus/backends/megatron/core/pipeline_parallel/zerobubble/scheduler/zbv.py]: better to remove commented code

File: primus/modules/trainer/megatron/trainer.py [hunk 5]
```diff
@@ -1788,30 +1849,49 @@ def train_step(
...
         args = get_args()
         timers = get_timers()
 
-        rerun_state_machine = get_rerun_state_machine()
-        while rerun_state_machine.should_run_forward_backward(data_iterator):
-            # Set grad to zero.
-            for model_chunk in model:
-                model_chunk.zero_grad_buffer()
-            optimizer.zero_grad()
+        def is_pipeline_stage_containing_loss():
```
Comment (lhzhang333) [inline primus/modules/trainer/megatron/trainer.py]: better to reuse `is_pipeline_stage_containing_loss` in `primus/backends/megatron/training/utils.py` or remove the redundant one

**Review by wenxie-amd — APPROVED** [review]
